### PR TITLE
XCM: Remote account converter

### PR DESCRIPTION
**Pull request by mustermeiszer**
Based on the discussion in #6612 this PR adds a converter to the codebase that allows to have a stable conversion from a local account (xcm sending location) to a remote account (xcm receiving location).

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2344314

**Comment by mustermeiszer** [comment]
@xlc @joepetrowski @jak-pan @girazoki WDYT about the converter like this? 

Happy to adapt to what is the compromise everybody is happy with.

**Comment by mustermeiszer** [comment]
Also @KiChjang do you already have some remarks?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2345685

**Comment by xlc** [comment]
I think a fixed prefix that includes parachain id could be useful. This acts as metadata and namespaced addresses. By only looking at an address, we will know it is an account belong to a certain parachain and impossible for one parachain to access account belong to another parachain

**Comment by mustermeiszer** [comment]
@xlc I thought about adding a fixed prefix to the resulting `[u8; 32]` and my main thoughts against it are
* a short prefix (e.g. `*b"P1000"`) &rarr; higher probability to have collisions with "local" accounts
* a larger prefix (e.g. `*b"ForeignP1000"`) &rarr; the entropy of the bytes is too low

Basically, these are just thoughts of someone who barely knows stuff about cryptography and the probabilities of collisions. But I think if we want a prefix we need to inspect this more thoroughly while we can trust the `blake2_256` output.

File: xcm/xcm-builder/src/location_conversion.rs [hunk 1]
```diff
@@ -22,6 +22,76 @@ use sp_std::{borrow::Borrow, marker::PhantomData};
...
+impl<AccountId: From<[u8; 32]> + Clone> Convert<MultiLocation, AccountId>
+	for ForeignChainAliasAccount<AccountId>
+{
+	fn convert_ref(location: impl Borrow<MultiLocation>) -> Result<AccountId, ()> {
+		let entropy = match location.borrow() {
+			// Used on the relay chain for sending paras that use 32 byte accounts
+			MultiLocation {
+				parents: 0,
+				interior: X2(Parachain(para_id), AccountId32 { id, .. }),
+			} => ForeignChainAliasAccount::<AccountId>::from_para_32(para_id, id),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/location_conversion.rs:45]:
```suggestion
			} => ForeignChainAliasAccount::<AccountId>::from_relay_32(para_id, id),
```

File: xcm/xcm-builder/src/location_conversion.rs [hunk 2]
```diff
@@ -22,6 +22,76 @@ use sp_std::{borrow::Borrow, marker::PhantomData};
...
+
+			// Used on the relay chain for sending paras that use 20 byte accounts
+			MultiLocation {
+				parents: 0,
+				interior: X2(Parachain(para_id), AccountKey20 { key, .. }),
+			} => ForeignChainAliasAccount::<AccountId>::from_para_20(para_id, key),
```
Comment (gavofyork) [inline xcm/xcm-builder/src/location_conversion.rs:51]:
```suggestion
			} => ForeignChainAliasAccount::<AccountId>::from_relay_20(para_id, key),
```

File: xcm/xcm-builder/src/location_conversion.rs (hunk 1 above)
Comment (mustermeiszer) [reply xcm/xcm-builder/src/location_conversion.rs:45]: This matching arms matches against messages received on the relay chain coming from a para-chain. I would argue that this should use the conversion from the parachain, allowing to have the same alias on each consuming chain of this converter.

File: xcm/xcm-builder/src/location_conversion.rs (hunk 2 above)
Comment (mustermeiszer) [reply xcm/xcm-builder/src/location_conversion.rs:51]: This matching arms matches against messages received on the relay chain coming from a para-chain. I would argue that this should use the conversion from the parachain, allowing to have the same alias on each consuming chain of this converter.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/multilocation-based-account-derivation/436/10

File: xcm/xcm-builder/src/location_conversion.rs (hunk 2 above)
Comment (gavofyork) [reply xcm/xcm-builder/src/location_conversion.rs:51]: Then I would add another item `from_child_20`, and with its own prefix.

File: xcm/xcm-builder/src/location_conversion.rs (hunk 1 above)
Comment (gavofyork) [reply xcm/xcm-builder/src/location_conversion.rs:45]: Then I would add another item `from_child_32`, and with its own prefix.

File: xcm/xcm-builder/src/location_conversion.rs [hunk 3]
```diff
@@ -22,6 +22,76 @@ use sp_std::{borrow::Borrow, marker::PhantomData};
...
+	}
+
+	fn reverse_ref(_: impl Borrow<AccountId>) -> Result<MultiLocation, ()> {
+		Err(())
+	}
+}
+
+impl<AccountId> ForeignChainAliasAccount<AccountId> {
+	fn from_para_32(para_id: &u32, id: &[u8; 32]) -> [u8; 32] {
+		(FOREIGN_CHAIN_PREFIX, para_id, id).using_encoded(blake2_256)
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply xcm/xcm-builder/src/location_conversion.rs:83]: Okay. I assumed that having a 32 bits difference between `from_relay_32` and `from_para_32` and a 256 bits difference between the `from_para_*` methods in the input stream would be enough difference to have sufficient bitflips with blake.

Will adapt then!

**Comment by mustermeiszer** [comment]
### RE: Methods
My reasoning for having only 
* `from_para_32`
* `from_para_20`
* `from_relay_32`

Is that the user will for a given Account `A` always know that his alias account `A'` is consistent across all chains using this converter. Making integrations with UI and wallets easier, where as when having the `from_child_*` this breaks this assumption for the relay chain. 

@gavofyork if having this is a blocker, I will adapt the PR.

**Comment by mustermeiszer** [comment]
@gavofyork or @KiChjang a last comment on the above and then we can probably finish this.

**Comment by gavofyork** [comment]
There may be unintended security consequences from having the same account ID used as the alias across multiple chains. I would certainly feel more confident if the aliases were different over each chain. That said, as long as it's strictly only for chains within the same security apparatus (i.e. not bridged locations), then it's probably fine to reuse the account alias.

**Comment by gavofyork** [comment]
This will break horribly with multi-level relay-chains: Account `A` on para index `P` controls account `h(A, P)` on all sibling, parent and child chains which use these primitives.

If you have multiple sets of parachains, then you'll naturally have index collisions. E.g.

```
         R
      /     \
    P1        P2
  /   \     /    \
P1.1 P1.2  P2.1 P2.2
```

Here, `P1.2` would be index 2 to `P1`, `P2.2` would be index 2 to `P2` and `P2` would be index 2 to `R`. Account `h(A, 2)` would therefore be controllable on `P2` by both `A` on `R`, and `A` on `P2.2`; this might be very problematic if you trust the logic of `R` much more than `P2.2`. 

This will never be a problem if you include the *relationship* within the hashed data.

**Comment by mustermeiszer** [comment]
> Account `h(A, 2)` would therefore be controllable on `P2` by both `A` on `R`, and `A` on `P2.2`; this might be very problematic if you trust the logic of `R` much more than `P2.2`.

The indices are wrong, but you are right:  

* `A` on `R` &rarr; controls `h(A)` on `P1` and `P2`
* `A` on `P1.2` &rarr; controls `h(A, 2)` on `P1` and `P1.1`
* `A` on `P2` &rarr; controls `h(A, 2)` on `R` and `P1`

&rarr; account `h(A, 2)` on `P1` is controllable by both `P2` and `P1.2`.

TBH I have not thought about multi-relay-chain setup. What do you mean with *relationship* in this context, how would that be extracted from the `MultiLocation`?

**Comment by gavofyork** [comment]
Yes; including the parents value is slightly more general, but either way would be fine.

**Comment by crystalin** [comment]
Hi @mustermeiszer, do you think you can do changes to adapt to the comments in the next days? We are waiting for this one to be merged before releasing a new runtime. If not, let me know and @girazoki can do the modifications
Thank you

**Comment by mustermeiszer** [comment]
@crystalin  Its updated. Can be reviewed and merged if all are happy.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2554023

**Comment by crystalin** [comment]
@mustermeiszer it seems to fail passing the test: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2554023#L1732

**Comment by mustermeiszer** [comment]
Docs did not format the tree correctly...

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2554784

**Comment by mustermeiszer** [comment]
@crystalin the two jobs failing seem not connected to this pr. cc @bkchr 

@gavofyork are you otherwise fine with the current state of the pr?

**Comment by mustermeiszer** [comment]
@gavofyork how do you feel about the current state? Can we merge it?

**Comment by girazoki** [comment]
@KiChjang @gavofyork is there anything else missing for this PR to be merged?

**Review by gavofyork — APPROVED** [review]
Looks good to me.

**Comment by mustermeiszer** [comment]
@KiChjang are you also happy with that? Can we merge that?

**Review by KiChjang — APPROVED** [review]
Ok for me, but needs an audit. Doesn't prevent this PR from landing however.

**Comment by mustermeiszer** [comment]
Cool. Let me know if you need anything for merging, but seems like it is up to you guys now.

**Comment by KiChjang** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/multichain-friendly-account-abstraction/1298/19

**Pull request merged**